### ISSUE TITLE
Treat type annotations in (generic) method and class signatures as API

### ISSUE THREAD

**Issue by vlsi** [autonomousapps/dependency-analysis-android-gradle-plugin#372]
**Is your feature request related to a problem? Please describe.**

It looks like the plugin does not always detect type annotations, so it suggests `api -> implementation` while the proper configuration should be `api`.

I do not have a standalone sample yet, and it looks like this issue results in false-positives for Caclite code: https://github.com/apache/calcite/pull/2362

**Describe the solution you'd like**

I expect that dependency-analysis should gather type annotations from all signatures (including [generic signatures](https://github.com/apache/calcite/blob/0fb14d553764f2a993ec56db4a36de2713ac1206/example/csv/src/main/java/org/apache/calcite/adapter/csv/CsvStreamScannableTable.java#L61)).

**Additional context**

The relevant bit of the report:

```
  Advice for project :example:csv
  Existing dependencies which should be modified to be as indicated:
  - implementation("org.checkerframework:checker-qual:3.10.0") (was api)
```

AFAIK, `checker-qual` is annotation-only dependency.

Sample uses for `@Nullable`. `buildHealth` suggests `implementation`, however, the annotations are used in the signatures of the public methods, so the dependency should be `api`.
https://github.com/apache/calcite/blob/0fb14d553764f2a993ec56db4a36de2713ac1206/example/csv/src/main/java/org/apache/calcite/adapter/csv/CsvTableScan.java#L89
https://github.com/apache/calcite/blob/0fb14d553764f2a993ec56db4a36de2713ac1206/example/csv/src/main/java/org/apache/calcite/adapter/csv/CsvTableFactory.java#L46

**Comment by autonomousapps** [comment]
Thanks for the report @vlsi. Can you tell me which version of the plugin you were using? [Version 0.68.0](https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/blob/main/CHANGELOG.md#version-0680) included a number of fixes relating to annotations.

If the issue persists, a minimal reproducer would help a lot.

**Comment by vlsi** [comment]
I used 0.71.0

**Comment by autonomousapps** [comment]
Thanks. I have an existing test suite covering various annotation cases, but I do not cover this scenario (yet).

**Comment by autonomousapps** [comment]
The fix for this will be in the next release.

**Comment by vlsi** [comment]
Nice, thank you.

**Comment by autonomousapps** [comment]
Just released 0.73.0 with this fix.

**Comment by vlsi** [comment]
Nice. It seems to heal all the issues with type annotations in Calcite.

**Comment by autonomousapps** [comment]
Woohoo! 

On May 5, 2021, GitHub ***@***.***> wrote:
> Nice. It seems to heal all the issues with type annotations in
> Calcite.
>

—
You are receiving this because you modified the open/close state.
Reply to this email directly, view it on GitHub
<https://github.com/autonomousapps/dependency-analysis-android-gradle-
plugin/issues/372#issuecomment-833328409>, or unsubscribe
<https://github.com/notifications/unsubscribe-
auth/ABJG5PIU4TF5LGY225KFOTDTMJF5HANCNFSM4YS6A3UQ>.
